azure/services/natgateways/natgateways.go

```diff
@@ -55,119 +57,61 @@ func New(scope NatGatewayScope) *Service {
 		if err != nil {
-			return errors.Wrapf(err, "failed to create nat gateway %s in resource group %s", natGatewaySpec.Name, s.Scope.ResourceGroup())
+			if !azure.IsOperationNotDoneError(err) || result == nil {
+				result = err
+			}
 		}
-		s.Scope.V(2).Info("successfully created nat gateway", "nat gateway", natGatewaySpec.Name)
-		natGateway := infrav1.NatGateway{
-			ID:   azure.NatGatewayID(s.Scope.SubscriptionID(), s.Scope.ResourceGroup(), natGatewaySpec.Name),
-			Name: natGatewaySpec.Name,
-			NatGatewayIP: infrav1.PublicIPSpec{
-				Name: natGatewaySpec.NatGatewayIP.Name,
-			},
+		if err == nil && result != nil {
+			// TODO: consider making OwnerResourceName() a no-op
+			subnetSpec := s.Scope.Subnet(natGatewaySpec.OwnerResourceName())
+			networkNatGateway, ok := natGateway.(network.NatGateway)
+			if !ok {
+				// TODO: maybe this error shouldn't take priority
+				result = errors.Errorf("created resource %T is not a network.NatGateway", natGateway)
+			}
+			// TODO: is it necessary to set the spec since it doesn't appear to change any behavior
+			subnetSpec.NatGateway.ID = *networkNatGateway.ID
+			s.Scope.SetSubnet(subnetSpec)
 		}
-		natGatewaySpec.Subnet.NatGateway = natGateway
-		s.Scope.SetSubnet(natGatewaySpec.Subnet)
 	}
-	return nil
-}
 
-func (s *Service) getExisting(ctx context.Context, spec azure.NatGatewaySpec) (*infrav1.NatGateway, error) {
-	existingNatGateway, err := s.Get(ctx, s.Scope.ResourceGroup(), spec.Name)
-	if err != nil {
-		return nil, err
-	}
-	// We must have a non-nil, non-"empty" PublicIPAddresses
-	if !(existingNatGateway.PublicIPAddresses != nil && len(*existingNatGateway.PublicIPAddresses) > 0) {
-		return nil, errors.Wrap(err, "failed to parse PublicIPAddresses")
-	}
-	// TODO do we want to eventually handle NatGateway resources w/ more than one public IP address?
-	// For now we assume the first one is the significant one
-	publicIPAddressID := to.String((*existingNatGateway.PublicIPAddresses)[0].ID)
-	resource, err := autorest.ParseResourceID(publicIPAddressID)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to parse Resource ID from PublicIPAddresses ID")
-	}
-	// We depend upon a non-empty ResourceName string
-	if resource.ResourceName == "" {
-		return nil, errors.Wrap(err, fmt.Sprintf("got unexpected ResourceName value from NatGateway PublicIpAddress, ResourceName=%s", resource.ResourceName))
-	}
-
-	return &infrav1.NatGateway{
-		ID:   to.String(existingNatGateway.ID),
-		Name: to.String(existingNatGateway.Name),
-		NatGatewayIP: infrav1.PublicIPSpec{
-			Name: resource.ResourceName,
-		},
-	}, nil
+	s.Scope.UpdatePutStatus(infrav1.NATGatewaysReadyCondition, serviceName, result)
+	return result
 }
 
 // Delete deletes the nat gateway with the provided name.
 func (s *Service) Delete(ctx context.Context) error {
-	ctx, _, done := tele.StartSpanWithLogger(ctx, "natgateways.Service.Delete")
-	defer done()
+	ctx, span := tele.Tracer().Start(ctx, "natgateways.Service.Delete")
```

Comment:
Is there any reason why you reverted `tele.StartSpanWithLogger`?

Reply:
I think that was just a mistake. Just to clarify, do we still want to use `tele.StartSpanWithLogger()` everwhere except for   `IsDone`, which is using `tele.Tracer().Start()` on all the async services so far.

Reply:
I think `IsDone` should be using `tele.StartSpanWithLogger()` too, not sure why it's not using it on the other services.